src/linear_algebra/matrix/transvection.lean

```diff
@@ -0,0 +1,754 @@
+`c` times the `j`-th line of `M` to its `i`-th line. Multiplying by it on the right corresponds
+to adding `c` times the `i`-th column to the `j`-th column. -/
+def transvection (c : R) : matrix n n R := 1 + c • E R i j
+
+@[simp] lemma transvection_zero : transvection i j (0 : R) = 1 :=
+by simp [transvection]
+
+section
+variable [fintype n]
+
+lemma update_row_eq_transvection (c : R) :
+  update_row (1 : matrix n n R) i (((1 : matrix n n R)) i + c • (1 : matrix n n R) j) =
+    transvection i j c :=
+begin
+  ext a b,
+  by_cases ha : i = a; by_cases hb : j = b,
+  { simp only [update_row, transvection, ha, hb, function.update_same, E_apply_one,
+      pi.add_apply, one_apply_eq, pi.smul_apply] },
+  { simp only [update_row, transvection, ha, hb, E_apply_zero, function.update_same,
+      pi.add_apply, ne.def, not_false_iff, pi.smul_apply, and_false, one_apply_ne] },
+  { simp only [update_row, transvection, ha, ne.symm ha, E_apply_zero, add_zero,
+      algebra.id.smul_eq_mul, function.update_noteq, ne.def, not_false_iff, dmatrix.add_apply,
+      pi.smul_apply, mul_zero, false_and] },
+  { simp only [update_row, transvection, ha, hb, ne.symm ha, E_apply_zero, add_zero,
+      algebra.id.smul_eq_mul, function.update_noteq, ne.def, not_false_iff, and_self,
+      dmatrix.add_apply, pi.smul_apply, mul_zero] }
+end
+
+lemma transvection_mul_transvection (h : i ≠ j) (c d : R) :
+  transvection i j c ⬝ transvection i j d = transvection i j (c + d) :=
+by simp [transvection, matrix.add_mul, matrix.mul_add, h, h.symm, add_smul, add_assoc]
+
+@[simp] lemma transvection_mul_apply (b : n) (c : R) (M : matrix n n R) :
+  (transvection i j c ⬝ M) i b = M i b + c * M j b :=
+by simp [transvection, matrix.add_mul]
+
+@[simp] lemma mul_transvection_apply (a : n) (c : R) (M : matrix n n R) :
+  (M ⬝ transvection i j c) a j = M a j + c * M a i :=
+by simp [transvection, matrix.mul_add]
+
+@[simp] lemma transvection_mul_apply_of_ne (a b : n) (ha : a ≠ i) (c : R) (M : matrix n n R) :
+  (transvection i j c ⬝ M) a b = M a b :=
+by simp [transvection, matrix.add_mul, ha]
+
+@[simp] lemma mul_transvection_apply_of_ne (a b : n) (hb : b ≠ j) (c : R) (M : matrix n n R) :
+  (M ⬝ transvection i j c) a b = M a b :=
+by simp [transvection, matrix.mul_add, hb]
+
+@[simp] lemma det_transvection_of_ne (h : i ≠ j) (c : R) : det (transvection i j c) = 1 :=
+by rw [← update_row_eq_transvection i j, det_update_row_add_smul_self _ h, det_one]
+
+end
+
+variables (R n)
+
+/-- A structure containing all the information from which one can build a nontrivial transvection.
+This structure is easier to manipulate than transvections as one has a direct access to all the
+relevant fields. -/
+structure transvection_struct :=
+(i j : n)
+(hij : i ≠ j)
+(c : R)
+
+noncomputable instance [nontrivial n] : inhabited (transvection_struct n R) :=
+by { choose x y hxy using exists_pair_ne n, exact ⟨⟨x, y, hxy, 0⟩⟩ }
```

Comment:
Having a non-computable `inhabited` instance seems pretty weird. I think it would be better to provide a `nonempty` instance and just `nolint` the request for an inhabited one.